### PR TITLE
[WIP]Schedule with backendV2

### PR DESCRIPTION
**Pull request by to24toro**
### Summary
This PR fixes the bug of scheduling with backendV2.
We aim to replace ScheduleConfig with Target when backendV1 will be deprecated in the future.

### Details and comments
- When making a `schedule`, `Target` and `ScheduleConfig` are prepared by dispatching the type of  backend. Now I consider `ScheduleConfig` class will be deprecated at the same time the backendV1 is dropped. 
- When deprecating `ScheduleConfig`, `Target` will become required and `ScheduleConfig` will be dropped.

### CODE REVIEW

File: qiskit/compiler/scheduler.py [hunk 1]
```diff
@@ -67,13 +68,22 @@ def schedule(
...
             dt = backend.dt
+        target = Target.from_configuration(
+            basis_gates=backend.operation_names,
+            num_qubits=backend.num_qubits,
+            coupling_map=backend.coupling_map,
+            inst_map=inst_map,
+            meas_map=meas_map,
+            dt=dt,
+        )
+        schedule_config = None
```
Comment (to24toro) [inline qiskit/compiler/scheduler.py:86]: A new target class is reconstructed from the provided backend parameters.

File: qiskit/compiler/scheduler.py [hunk 2]
```diff
@@ -34,7 +36,11 @@ def _log_schedule_time(start_time, end_time):
     log_msg = "Total Scheduling Time - %.5f (ms)" % ((end_time - start_time) * 1000)
     logger.info(log_msg)
 
-
+@deprecate_arg(
+    "backend",
+    additional_msg=("'backend' argument becomes required from Optional."),
+    since="0.25.0",
+)
```
Comment (to24toro) [inline qiskit/compiler/scheduler.py:43]: Target is made by some backend parameters and I would like to make backend required from Optional.

Comment (TsafrirA) [reply qiskit/compiler/scheduler.py:43]: @to24toro 
I think this is an incorrect use of `@deprecate_arg`. The warning will only be raised when the user *does* provide the argument explicitly. This is the opposite of what you want - you want the warning to be raised when the user doesn't provide it.
I would have placed the warning in the function itself:
```
if backend is None:
  # raise warning
```

File: qiskit/compiler/scheduler.py [hunk 3]
```diff
@@ -65,15 +67,30 @@ def schedule(
...
         if dt is None:
             dt = backend.dt
+        target = Target.from_configuration(
+            basis_gates=backend.operation_names,
+            num_qubits=backend.num_qubits,
+            coupling_map=backend.coupling_map,
+            inst_map=inst_map,
+            meas_map=meas_map,
+            dt=dt,
+        )
```
Comment (to24toro) [inline qiskit/compiler/scheduler.py:92]: `from_configuration` requires basis_gates. 
How do we create target from configuration when the backend is None?

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]
Thanks @to24toro this is a great start of this final piece of modernization in the pulse stack. The main purpose of this PR is not only adding target to the user API, but also for completely removing the internal use of inst_map. Note that inst_map can be replaced with `Target.get_calibration`, and a minimal target required by scheduling can be created with

```python
target = Target(meas_map=meas_map)
target.update_from_instruction_schedule_map(inst_map)
```

so you can eliminate every internal use of inst_map without API break. `ScheduleConfig` is no longer necessary. 

I also think `Schedule` is a legacy format and should be deprecated at some point. So this scheduler must return `ScheduleBlock` rather than `Schedule`. Of course this is an API break, and should not change return object in this PR. But in the sense of internal cleanup, you can build a `ScheduleBlock` first and return `Schedule` with `target_qobj_transform` for backward compatibility. In this way you can delegate cumbersome t0 computation to the block context alignment.

In the followup, you can add new argument

```python
schedule(..., return_block=False)
```

with deprecation warning, i.e. to let user set `return_block=True`, and eventually drop `target_qobj_transform` to always return `ScheduleBlock`. To ease this future refactoring, I think it's worth changing the scheduling logic in this PR as a part of cleanup.

Here is a toy code to build a schedule from block and target.

```python
from qiskit.providers.fake_provider import FakeHanoiV2
from qiskit import circuit, transpile
from qiskit.pulse import ScheduleBlock, Schedule
from qiskit.pulse.transforms import AlignLeft, AlignSequential, target_qobj_transform

from qiskit.pulse.builder import _PulseBuilder

backend = FakeHanoiV2()
target = backend.target

circ = circuit.QuantumCircuit(2, 2)
circ.x(0)
circ.cx(0, 1)
circ.x(1)
circ.sx(0)
circ.measure(0, 0)
circ.measure(1, 1)

# Example of ASAP scheduling
block = ScheduleBlock(circ.name, circ.metadata, alignment_context=AlignSequential())
sub_block = ScheduleBlock(alignment_context=AlignLeft())

qubits_in_context = set()
for inst in circ.data:
    qargs = tuple(map(lambda q: circ.find_bit(q).index, inst.qubits))
    
    if set.intersection(qubits_in_context, set(qargs)):
        # emit new prallel context
        block.append(sub_block, inplace=True)        
        sub_block = ScheduleBlock(alignment_context=AlignLeft())
        qubits_in_context.clear()   
        
    calibration = target.get_calibration(inst.operation.name, qargs, *inst.operation.params)
    if isinstance(calibration, Schedule):
        # this is a temporary hack.
        # currently backend calibration is reported in Schedule
        calibration = _PulseBuilder._naive_typecast_schedule(calibration)
    sub_block.append(calibration, inplace=True)
    qubits_in_context |= set(qargs)

if len(sub_block) != 0:
    block.append(sub_block, inplace=True)
    
schedule = target_qobj_transform(block)
```

Note that meas map constraint is not considered at all. So you need to add more lines. However, as you can see, the logic looks much cleaner.

**Comment by to24toro** [comment]
I updated a schedule logic at https://github.com/Qiskit/qiskit-terra/pull/10285/commits/bc421ad1c732dd4a86f9e6afd3156bc65fe4a3df and a sequence logic at https://github.com/Qiskit/qiskit-terra/pull/10285/commits/420596002edd30b7807547ffbc2bec5dc375ec58.

**Comment by to24toro** [comment]
I will reopen in another PR after reorganize it.

**Comment by to24toro** [comment]
I close this PR and reopened at https://github.com/Qiskit/qiskit-terra/pull/10564

**Pull request closed without merging**